### ISSUE TITLE
Backgrounds should extend across the whole terminal

### ISSUE THREAD

**Issue by brandur** [jwilm/alacritty#504]
Background colors should stretch across the entire terminal rather than be constrained to the last character.

Here's Alacritty:

<img width="557" alt="screenshot 2017-04-02 09 18 14" src="https://cloud.githubusercontent.com/assets/96890/24588886/c47cf73c-1785-11e7-93a9-b2c451754049.png">

And here's iTerm2 for comparison:

<img width="635" alt="screen shot 2017-04-02 at 9 22 07 am" src="https://cloud.githubusercontent.com/assets/96890/24588891/e4692afc-1785-11e7-83db-1cad00b0adde.png">

**Comment by boxofrox** [comment]
A console program that wants to highlight an entire line should pad the line with spaces when it prints to explicitly apply the background color to all cells in the line.

If the terminal does this automatically, how do other programs that only want to highlight part of a line stop the terminal from carrying the background color to the end of the line?

What program are you using in your screenshots?

**Comment by brandur** [comment]
> What program are you using in your screenshots?

This is the email client [Mutt](http://www.mutt.org/), an email client so long-lived, so well-established, and so well-respected, that whatever behavior it expects out of its host terminal will probably need to be reproduced out of pure practicality alone.

I should also note here that I may have the diagnosis wrong. You'd obviously only expect a background color change to last as long as its surrounding escape codes, but this may be some alternate curses-style control that does the entire line. I'll try to dig into its implementation a bit and see if I can figure out what it's doing.

**Comment by boxofrox** [comment]
The issue appears to affect MacOS.  On ArchLinux, mutt in alacritty behaves like your iTerm2 example.

![image](https://cloud.githubusercontent.com/assets/636691/24593118/1183cda4-17ee-11e7-816e-a0b36cb2c9e7.png)

**Comment by boxofrox** [comment]
For any devs that need steps to reproduce:

1.  Install `mutt`.  I assume `brew install mutt` will work, @brandur?
2.  Grab a mailing list archive to view in `mutt`\[[1]].
    1.  Download the gzipped archive of a mailing list.
    `curl -LO http://openvswitch.org/pipermail/dev/2016-June.txt.gz`
    2.  Unzip the archive.  `gunzip 2016-June.txt.gz`
    3.  Open the archive in `mutt`.  `mutt -f 2016-June.txt`.
3.  Use Arrow keys to move up/down list.  Use `o` then `t` to sort threads.
4.  Observe whether highlight extends to right edge of terminal.

[1]: https://that.guru/blog/mutt-archives/
\[[1]]: https://that.guru/blog/mutt-archives/

**Comment by brandur** [comment]
Thanks @boxofrox! Those instructions look roughly right to me.

**Comment by brandur** [comment]
And just as a note: I'm happy to tackle a fix here, and especially if someone could point me in the right direction.

I suspect that the issue probably stems from something like a "background highlight" control code combined with something like a "until end of line" control code not coming properly, but I don't know enough about how this stuff works right now to say for sure. I'll try to dig into something like curses and see if I can see what's going on from there.

**Comment by jwilm** [comment]
Hey @brandur and @boxofrox,

Thanks for the report and discussion here. This turns out to be an issue of Alacritty claiming to support [BCE (back color erase)](http://invisible-island.net/ncurses/ncurses.faq.html#bce_mismatches) but not actually supporting it. Closing as a duplicate of #85, #195.

@brandur I've got a PR up which implements bce, but it broke a test that I haven't had time to investigate: #407.
